### ISSUE TITLE
Catching exception Apartment::TenantNotFound

### ISSUE THREAD

**Issue by shakthimaan** [influitive/apartment#235]
Hi,

I am using Rails 3.2.21 and apartment 1.0.1 with the following config/initializers/apartment.rb. 

```
require 'apartment/elevators/generic'

Apartment.configure do |config|
  config.tenant_names = ['alpha', 'beta', 'gamma']
end

Rails.application.config.middleware.use 'Apartment::Elevators::Generic', lambda { |request|
  request.env['HTTP_X_TENANT']                                                                           
} 
```

It works fine when the tenant exists in the MySQL database server. When there is no tenant, it throws an Apartment::TenantNotFound exception. How can I catch it?

#1 I saw the following response, but, .prepend is not available in Rails 3.2:
http://stackoverflow.com/questions/27188411/apartment-ruby-gem-want-to-catch-an-exception

#2 I also tried adding the following code in application_controller.rb, but, it didn't get triggered:

```
class ApplicationController < ActionController::Base
  respond_to :json

  rescue_from Apartment::TenantNotFound do |exception|                                                                                                                                                              
    render :json => { :errors => "Tenant not found!" }                                                                                                                                                              
  end
   
end
```

Any suggestions?

**Comment by bradrobertson** [comment]
FYI [Module::prepend](http://ruby-doc.org/core-2.0.0/Module.html#method-i-prepend) is a ruby 2.0 feature, nothing to do with Rails.

You can write your own [middleware](https://github.com/influitive/apartment#switching-tenants-per-request) that extends Apartment's middleware to rescue exceptions. I *want* to be able to hook apartment into the Rails stack so you can `rescue_from` as you've done above, but right now this is NOT supported. I don't actually really know how to do this. The issue is that Apartment gets hooked into the Rack stack further down (outside of controllers) so it's hard to tap into.

Basically if you write your own middleware and rescue that exception, you can do whatever you want in a Rack response. Something like this:

```ruby
class MyMiddleware < ::Apartment::Elevators::Subdomain
  def call(env)
    super
  rescue ::Apartment::TenantNotFound
    [301, {'Location' => 'http://some.url'}, ['redirect']]
  end
end
```

And use that instead of `Apartment::Elevators::Generic`

I've been meaning to add a section in the README since this comes up a fair bit but haven't had time. If you get this working I'd *love* a PR that adds it to the docs. It'd help out a lot.

**Comment by shakthimaan** [comment]
I got it working with the following code snippets.

config/initializers/apartment.rb
```
require 'apartment/elevators/subdomain'
require 'rescued_apartment_middleware'

Apartment.configure do |config|
  config.tenant_names = ['alpha', 'beta', 'gamma']
end

Rails.application.config.middleware.use(TenantMiddleware)
```

In lib/rescued_apartment_middleware.rb:
```
class TenantMiddleware < Apartment::Elevators::Subdomain
  def call(*args)
    begin
      super
    rescue Apartment::TenantNotFound
      Rails.logger.error "ERROR: Apartment Tenant not found: #{Apartment::Tenant.current.inspect}"
      return [404, {"Content-Type" => "application/json"}, ["Error" => "Tenant not found"]]
    end
  end

  def parse_tenant_name(request)
    request.env['HTTP_X_TENANT']
  end
end
```

**Comment by paulomcnally** [comment]
<img width="659" alt="captura de pantalla 2019-01-09 a la s 10 30 55 p m" src="https://user-images.githubusercontent.com/181533/50946323-41818780-145e-11e9-9af2-54c625d06cbf.png">

**Comment by paulomcnally** [comment]
Fixed with:

```
# frozen_string_literal: true

module RescueApartmentMiddleware
  def call(*args)
    begin
      super
    rescue Apartment::TenantNotFound, TypeError => error
      puts error
      return [404, { "Content-Type" => "text/html" }, ["#{ File.read(Rails.root.to_s + '/public/404.tenant.html') }"]]
    end
  end
end

```

**Comment by archonic** [comment]
I've found overriding the Generic class and then ensuring return of `@app.call(env)` to be the best way to rescuing Apartment::TenantNotFound in a way where Rails can serve the response.

At the top of config/initializers/apartment.rb:
```
require "rescued_apartment_elevator"
```

lib/rescued_apartment_elevator.rb:
```
# frozen_string_literal: true

require 'rack/request'
require 'apartment/tenant'

module Apartment
  module Elevators
    class Generic
      def call(env)
        request = Rack::Request.new(env)
        database = @processor.call(request)
        if database
          Apartment::Tenant.switch(database) { @app.call(env) }
        else
          @app.call(env)
        end
      rescue Apartment::TenantNotFound => error
        Rails.logger.error "ERROR: Apartment Tenant or Schema not found: #{error}"
        @app.call(env)
      end
    end
  end
end
```

In application_controller.rb:
```
rescue_from ActionController::RoutingError,
    ActiveRecord::RecordNotFound,
    Apartment::TenantNotFound,
    with: :render_404

...
private

def render_404(_exception)
  render "errors/not_found"
end
```
